### PR TITLE
Add a flag to IO types to skip fsgrid fg DRO output.

### LINKED ISSUE #692: Writing fg_variables to a separate output file
Currently including `fg_`variables in bulk outputs gets quite expensive for large 3D simulations. With increasing levels of AMR, this will become unfeasible to include in every output file.

Vlasiator already includes the possibility to declare several different output types with different time cadences, but these only differ for VDF strides, not included variables. I suggest we add a toggle for each type to include `fg_`, `vg_`, and `ig_` variables, defaulting to on for others and off for fg_. 

Flags to be added along the lines of 
```
RP::addComposing("io.system_write_distribution_stride", "Every this many cells write out their velocity space. 0 is none. [Define for all groups.]");
```
close to https://github.com/fmihpc/vlasiator/blob/5e656cf2e4e2dcc1fd4d521010fa655ef1b2fa09/parameters.cpp#L191

And then in iowrite we should pass the filedescriptor index to https://github.com/fmihpc/vlasiator/blob/5e656cf2e4e2dcc1fd4d521010fa655ef1b2fa09/iowrite.cpp#L1432 and check within each writeDataReducer() based on the name whether to write or return.

### PR DESCRIPTION
**Pull request by ykempf**
Fixes #692. Comments?

### CODE REVIEW

**Comment by ykempf** [comment]
This means adding that line to all active cfgs. Heh so to testpackage cfgs as soon as there is an agreement on the principle of this I suppose.

**Comment by ykempf** [comment]
I rean a test with restarts and it had fg_E, fg_PERB in there all right and did not complain about anything upon restarting.

**Comment by ursg** [comment]
So, functionally and technically good.
I'm just wondering if I can come up with any situations where this would bite us in the backside. Apart from the obvious change that testpackage now requires, I'm wondering how this complicates matters for people wanting to analyze simulation data afterwards: if you want any expressions involving both vg_ and fg_ variables, you'd then have to make sure to fish them from separate files correctly.
But if you want to do that, you already have to carefully interpolate and match up the grids, so maybe that's not so bad?

**Comment by ykempf** [comment]
Part of the reason was my push to not save fg at all (as was done, maybe at odds with announced things/spreadsheets we didn't go through in december) in production. It's a lot easier (crunching, RAM, coordinates, grid (mis)matching) to have all residing on vg for analysis. So, yes, in case one wants to analyse including fg they'll have to pay attention. But the files containing fg will contain everything, i.e. a duplicate of the vg variables. Those are cheaper. And then we can set either no VDFs in the fg files, or more than in regular files as we used to do with distributions files.

**Comment by markusbattarbee** [comment]
Indeed, this'll at least allow us to save fs grids at some lower cadence, and then perhaps also with more VDFs.

I went through the code and although it's hampered a bit by the limitations of the way boost handles these multi-element values, it's as neat as can be hoped for, and I think should serve it's purpose well. 

Perhaps, if I may suggest, there could be a check that in parameters.cpp at line 544 or so, that if `P::systemWriteFsGrid.size()==0`, then a vector of length `maxSize` is initialized there with values true, making it in effect the default value?

Another interesting note here is that I don't see an ionospheric grid passed to `writeGrid()`... ;) so I guess it just suitably materializes from somewhere. Harrumpf.

**Comment by ykempf** [comment]
Like so?

**Review by markusbattarbee — APPROVED** [review]

**Comment by ursg** [comment]
> Another interesting note here is that I don't see an ionospheric grid passed to `writeGrid()`... ;) so I guess it just suitably materializes from somewhere. Harrumpf.

Yeah, well, that is a global member of the SBC namespace. Not to elegant in terms of proper software design, I admit...

File: iowrite.cpp [hunk 1]
```diff
@@ -318,13 +318,21 @@ bool writeDataReducer(const dccrg::Dccrg<SpatialCell,dccrg::Cartesian_Geometry>&
...
+                      cint dataReducerIndex,
                       Writer& vlsvWriter){
    map<string,string> attribs;
    string variableName,dataType,unitString,unitStringLaTeX, variableStringLaTeX, unitConversionFactor;
    bool success=true;
-
+   
+   if(!writeFsGrid) { // if we shouldn't write fsgrid DROs
+      variableName = dataReducer.getName(dataReducerIndex);
+      if(variableName.find("fg_", 0) == 0) { // and if the DRO's name includes the string "fg_"
```
[the remark this answers is not in the document]
Comment (ykempf) [reply iowrite.cpp:331]: And uhm I don't know what you are talking about. ;)

File: iowrite.cpp [hunk 2]
```diff
@@ -1298,8 +1306,8 @@ bool writeGrid(
    DataReducer* dataReducer,
    const uint& index,
```
[the remark this answers is not in the document]
Comment (ykempf) [reply iowrite.cpp:1307]: OK done.

File: parameters.cpp [hunk 3]
```diff
@@ -537,6 +541,18 @@ void Parameters::getParameters() {
...
       }
    }
+   if (P::systemWriteFsGrid.size() != maxSize) {
+      if (P::systemWriteFsGrid.size() == 0) {
+         for (uint i = 0; i < maxSize; i++) {
+            P::systemWriteFsGrid.push_back(true);
+         }
+      } else {
+         if (myRank == MASTER_RANK) {
+            cerr << "ERROR io.system_write_fsgrid_variables should be defined for all file types." << endl;
```
[the remark this answers is not in the document]
Comment (ykempf) [reply parameters.cpp]: yup

**Comment by ykempf** [comment]
Here we go I guess?

**Pull request merged**